test/cpp/api/serialize.cpp

```diff
@@ -553,6 +558,56 @@ TEST(SerializeTest, Optim_SGD) {
   is_optimizer_state_equal<SGDParamState>(optim1.state(), optim1_2.state());
 }
 
+TEST(SerializeTest, Optim_Adam) {
+  test_serialize_optimizer<Adam, AdamOptions, AdamParamState>(AdamOptions().amsgrad(true).weight_decay(0.5).betas(std::make_tuple(0.6, 0.999)));
```

Comment:
there's a test failure at this line: https://github.com/pytorch/pytorch/blob/1546d2afeb98fcd7bc5b58261d6e31ad7794e833/test/cpp/api/serialize.cpp#L154 any idea why @vincentqb 
p.s. the step value is being incremented correctly

Reply:
P.S. we have a similar check for optimizers in Python

Reply:
For reference, from [here](https://app.circleci.com/jobs/github/pytorch/pytorch/4697533):
```
[ RUN      ] SerializeTest.Optim_Adam
step1 model1
step2 model1
step1 model2
step1 model2
step1 model3
..\test\cpp\api\serialize.cpp(159): error: Value of: param1[name].norm().item<float>() != param2[name].norm().item<float>()
  Actual: false
Expected: true
[  FAILED  ] SerializeTest.Optim_Adam (6 ms)
```

Can you link to the equivalent in python?

Reply:
Do you know which commit in this PR caused this?

Reply:
resolved: optimizer wasn't taking big enough step and the value exp_avg/denom * - step_size,  was coming out to be very close for model1 step2 and model2 step1 (with optim2_2).  step_size = options.lr() / bias_correction1 and bias_correction1 is a function of step.val(), hence increasing the lr increased the step_size in both cases but by different amounts and hence increased the difference in values in the two cases.